src/http-client.js

```diff
@@ -104,14 +120,26 @@ export class HttpClient {
   fetch(input: Request|string, init?: RequestInit): Promise<Response> {
     this::trackRequestStart();
 
-    let request = Promise.resolve().then(() => this::buildRequest(input, init, this.defaults));
+    let request = Promise.resolve().then(() => this::buildRequest(input, init, this.defaults))
+      .then( result => {
+        const { retryConfig } = result;
+        if (retryConfig) {
+          result.retryConfig = retryConfig;
+        }
+        return result;
+      });
+
     let promise = processRequest(request, this.interceptors)
       .then(result => {
         let response = null;
 
         if (Response.prototype.isPrototypeOf(result)) {
           response = result;
         } else if (Request.prototype.isPrototypeOf(result)) {
+          if (result.retryConfig) {
+            result.retryConfig.requestClone = result.clone();
+            result.retryConfig.client = this;
+          }
           request = Promise.resolve(result);
           response = fetch(result);
```

Comment:
@AshleyGrant I think it might be possible to enhance this section to behave the way @CasiOo originally believed it to behave. That is, if the result of fetch is a `Request`, then resolve `fetch(result)`. That would open up a lot of doors including (but not limited to) a fully decoupled retry interceptor.

If possible, I think we should make every effort to go down that path. It is a more robust implementation that will be better for the fetch client in the long run. What do you think?